packages/block-editor/src/components/use-editor-fonts-resolver/index.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useMemo, useCallback } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { getDisplaySrcFromFontFace, loadFontFaceInBrowser } from './utils';
+import { store as editorStore } from '../../store';
+
+function useEditorFontsResolver() {
+	const [ loadedFontUrls, setLoadedFontUrls ] = useState( new Set() );
+
+	const { currentTheme = {}, fontFamilies = [] } = useSelect( ( select ) => {
+		return {
+			currentTheme:
+				// Disable Reason: Using 'core' as string to avoid circular dependency importing from @wordpress/core-data.
+				// eslint-disable-next-line @wordpress/data-no-store-string-literals
+				select( 'core' )?.getCurrentTheme(),
```

Comment:
To get the base url of the theme's fonts.

Reply:
> to "resolve" the files in the server

Do you mean when the `theme.json` data is read from the file / database?

> add a new _unstableType for the "styles" object received by the editor.

Sotty, but I don't understand this.

Reply:
> Do you mean when the theme.json data is read from the file / database?

Yes, basically in this place, in the code base. We read global styles and enqueue "styles" setting in the block editor. We could add a new "style" in that array with a specific `"_unstableType" => "font"` that we treat in a custom way in the `EditorStyles` component. 

https://github.com/WordPress/wordpress-develop/blob/bf09fe506620678fb82c3b872309edda0ed8ce61/src/wp-includes/block-editor.php#L497-L553

I know that function is in WordPress Core but in Gutenberg we have a hook to override the output of this function `block_editor_settings_all` filter I think.

---

I know that we used this approach in the past for "svg filters" which are also "special styles" that are not just CSS. @ajlende would know more but I think we moved a little bit from it (not entirely sure why though)

Reply:
Thanks for adding context. The problem I see with this approach is that the backend function is not called when the client-side changes are applied. So, if the user is switching style variations, that is done entirely in the client (as far as I understand), and no Rest API request is sent to the backend. Would this approach be useful in that case?

Reply:
>  The problem I see with this approach is that the backend function is not called when the client-side changes are applied

This is not a problem, and we already address this in the "GlobalStylesRenderer" like I said [here](https://github.com/WordPress/gutenberg/pull/65019#discussion_r1754975668). We should just make sure the resolution of the fonts is also done within that `useGlobalStylesOutput` hook.

Reply:
Let's say we add a a json array of font-families in the backed using `gutenberg_get_block_editor_settings`, we process those families font faces to include the right URLs if that's needed. I get that. If the user switches the content of the fontFamilies theme.json data with data coming from a style variation those families will still have theme relative (`file:/...`) URLs. I don't see how that can be handled in the code [linked](https://github.com/WordPress/gutenberg/pull/65019#discussion_r1754975668).

Reply:
you're right in the sense that `useGlobalStylesOutput` is also a block-editor API it seems. I thought personally that it was an "edit-site" API, so we could fetch the theme and do the resolution from there. That said it seems `useGlobalStylesOutput` is a private API for now. I think we can add the "theme path/url" as an argument to address that.

It's unfortunate thought that "resolving fonts" is the only thing from a "theme.json" object that requires outside data (theme) in order to resolve the styles properly.

Reply:
This touches upon many of the issues we grappled with when dealing with background image relative paths.

Ultimately, we settled on a balance between the server and client, where:

- the server resolves relative paths
- sends them to the client in the `_links` array in the REST response
- the client then has access to both relative and absolute paths for different tasks, e.g., using relative paths for exporting theme.json, and absolute paths for building CSS

If you think something similar would be appropriate here, I'd be happy to help out.

See the original PR: https://github.com/WordPress/gutenberg/pull/61271

Also, there is a _slightly_ related discussion about whether absolute paths to theme assets should be saved in the global styles CPT at all: https://github.com/WordPress/gutenberg/issues/61809

Reply:
Thanks @ramonjd that is indeed very similar. It seems like we're starting to have multiple needs for the same thing: resolving relative paths during "global styles resolution". It would be good to normalize/document/generalize this a bit so the path is clear if it comes up again.

Reply:
Thanks for the additional context. I took a look to the `_links` implementation and (at a first glance) it seems to be more complex than needed for this particular dutie (resolve the theme relative path of a font face in the browser). Does the `_links`knowledge of global styles implies that the global styles components know about the rest API design?  It's good the intention to provide a common way to do this but is the backend code added really needed? 

Could injecting the currentTheme into the editor settings using the GlobaStylesRenderer component be a simpler approach that works? I tried that in this commit: https://github.com/WordPress/gutenberg/pull/65019/commits/8ced871bc797e8b4d3819892e01ea7133002938f

https://github.com/WordPress/gutenberg/blob/8ced871bc797e8b4d3819892e01ea7133002938f/packages/edit-site/src/components/global-styles-renderer/index.js#L18-L30

`GlobaStylesRenderer` is inside the `edit-site` package and I imagine that is ok it knows about a theme, is this correct?


Maybe the problem is that we would need to read the theme from the editor settings in the useEditorFontsResolver which is inside `block-editor` package.

https://github.com/WordPress/gutenberg/blob/8ced871bc797e8b4d3819892e01ea7133002938f/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L13-L25

How bad is this solution?

Reply:
> resolve the theme relative path of a font face in the browser

What does "resolve" mean in this context? Resolving to get the full URL I presume.

Can you be sure that `currentTheme.stylesheet_uri` is the [base URL](https://github.com/WordPress/gutenberg/pull/65019/files#diff-1db5cd8aa6c0226c4bf0053532d6419d41d6c477b6b9360633c2595a3a49e883R43) of the font assets always?

If so, then it might be enough to do it on the clientside.

One of the considerations with which I had to contend when looking at theme.json image assets is that the file could be in the current theme's directory, or, in the case of a child theme, inherited by the parent theme.

Also there are [filters](https://developer.wordpress.org/reference/hooks/stylesheet_uri/) that allow themes/plugins etc to change the stylesheet URI.

Reply:
I'm trying to follow the pattern of [this PR](https://github.com/WordPress/gutenberg/pull/61271) using the _link array. It seems to work with the regular theme.json, but I'm unable to access the links for the style variations.

Reply:
I'm trying to understand the use case. What do we need to do with the resolved font paths? Preload the fonts? Could you provide some testing steps and expectations for style variations? 

Just to confirm, if I run `await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentThemeGlobalStylesVariations()` in the console I see that the variations' fonts are resolved in the `_links` object.

> How I can pass the variation _settings to the editor settings?

Aren't the links already [available in the settings](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/editor/src/components/provider/use-block-editor-settings.js#L289-L289)?

It's private, so, in the [useEditorFontsResolver](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L14-L14) `useSelect` you'd have to import the private symbol key (`import { globalStylesLinksDataKey } from '../../store/private-keys';`) and access it like `settings[ globalStylesLinksDataKey ]`.

I just logged the settings out in that hook and I can see the font `_links` object, and when I change variations the links update.

I'm not sure that helps, maybe it leads you in the right direction?


------

Just to clarify and contrast what I was working on with background images, the resolved background image paths are used to build the correct CSS. It's done at this point in the global styles output generation:

https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L308-L314

`tree` contains the embedded `_links` object and path resolution is done "on the fly" when outputting CSS.

Reply:
> What do we need to do with the resolved font paths? Preload the fonts? Could you provide some testing steps and expectations for style variations?

This is the issue we are trying to fix: https://github.com/WordPress/gutenberg/issues/59965
(load the fonts in the main frame and in the editor iframes when the fonts are defined in variations)

Reply:
In this [commit](https://github.com/WordPress/gutenberg/pull/65019/commits/146198902f9e68f23681157bcafe4497143f6cdb) I added the JS consumption of the _links. It seems to work for the main editor but not for the editors loaded to display the style variations.



The commit includes a console.log to display the font families and the _links.

https://github.com/WordPress/gutenberg/blob/146198902f9e68f23681157bcafe4497143f6cdb/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L22-L34

In the screencast you can see how the editors instances created to display the style variations doesn´t get the right list of font families (the ones from the style variation) and the _links array is empty.

[Screencast from 27-09-24 14:04:29.webm](https://github.com/user-attachments/assets/a7c4bc97-78ba-46d2-909b-af872ea9dacf)

Is there any way to get the right list of font families and _links in the editors used to feature a style variation?

Reply:
So it looks like here you need all the variations' fonts at once for the sidebar. Is that right?

If you look at where the [key value is set in the settings](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/editor/src/components/provider/use-block-editor-settings.js#L200), you'll notice that the `_links` passed to settings contains only the resolved links for the merged global styles that are currently loaded in the editor. 

That's why the resolved paths for style variations only work once you click on them, because it's it that point they're loaded into the editor.

So you'll probably need to iterate over the output of `await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentThemeGlobalStylesVariations()` as hinted at above since they contain all the variations their resolved paths, and somehow preload the fonts for each variation.

However you can't use these core selectors in the block editor package, that is, in `useEditorFontsResolver()`.

Rather, you might like to tackle from the edit-site package, where the [style variations are rendered](https://github.com/WordPress/gutenberg/blob/146198902f9e68f23681157bcafe4497143f6cdb/packages/edit-site/src/components/global-styles/style-variations-container.js#L124-L124).

For example, export `useEditorFontsResolver()` as a private method and have it accept `_links` and `fontFamilies` as dependencies. I don't know if that will work, but it's the first thing I'd try without having extra context.